### PR TITLE
added pytorch simulation example

### PR DESCRIPTION
**Pull request by MakGulati**
This involves partitioning of cifar data using numpy and training a simple neural network in pytorch similar to tf example.

### CODE REVIEW

File: examples/simulation_pytorch/pytorch_simulation.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+def run_simulation(num_rounds: int, num_clients: int, fraction_fit: float):
+    """Start a FL simulation."""
+    logging.basicConfig(
+        filename="pytorch_simulation.log",
+        filemode="w",
+        level=logging.DEBUG,
+        format="%(asctime)s %(levelname)s: %(message)s",
+        datefmt="%d-%m-%Y %I:%M:%S %p",
+    )
```
Comment (tanertopal) [inline examples/simulation_pytorch/pytorch_simulation.py:98]:
```suggestion
        fl.common.logger.configure("simulation", "simulation.log")
```

**Review by tanertopal — COMMENTED** [review]
@MakGulati your PR is very much appreciated. It already looks great. I made one suggestion in the code and here a few more:

1. Could you run the ./dev/format.sh script once and commit the new formatted code? Unfortunately, we are not checking this yet in the CI but will change that soon.
1. To keep the naming overall consistent with other examples would you mind renaming the files:
    1. `pytorch_dataloader` => `dataset.py`
    1. `pytorch_simulation.py` => `simulation.py`
1.  Add a `pyproject.toml`
1. Add a `README.md` (similar to the e.g. other simulation example). The idea is that each example is also a boilerplate which can easily be cloned as described in the other examples `README.md` files

File: examples/simulation_pytorch/pytorch_simulation.py (hunk 1 above)
Comment (tanertopal) [reply examples/simulation_pytorch/pytorch_simulation.py:98]: @MakGulati I tried running your example and it gets stuck after the sampling. Would it be possible for you to also add a Dockerfile and `run.sh` script similar to the other simulation example in which the training finishes? I want to make the issue lies is not because of different environment settings.

Comment (MakGulati) [reply examples/simulation_pytorch/pytorch_simulation.py]: @tanertopal I updated the code as per your advice but I am still facing some issues(memory overflow) while running docker container. I was using conda environment on my machine. Could you please try this conda environment settings from my [environment.yml ](https://gist.github.com/MakGulati/4f32a98924adbee467332387f644b21b).
Pleas let me know if there is still some problem. Meanwhile I will check what I can do with the docker file.

Comment (MakGulati) [reply examples/simulation_pytorch/pytorch_simulation.py]: @tanertopal Please see the updates. I hope I covered all the requirements.
Thank you :)

Comment (tanertopal) [reply examples/simulation_pytorch/pytorch_simulation.py]: @MakGulati amazing thank you too! I will try it out soon and provide you with feedback.

File: examples/simulation_pytorch/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+This will create a new directory called `simulation_pytorch` containing the following files:
+
+```shell
+-- dataset.py
+-- Dockerfile
+-- pyproject.toml
+-- README.md
+-- run.sh
+-- simulation.py
```
Comment (tanertopal) [inline examples/simulation_pytorch/README.md:35]:
```suggestion
-- SimpleNet.py
-- simulation.py
```

File: examples/simulation_pytorch/simulation.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        processes.append(client_process)
+
+    # Block until all processes are finished
+    for p in processes:
+        p.join()
+
+
+if __name__ == "__main__":
+    torch.multiprocessing.set_start_method("spawn") 
+    run_simulation(num_rounds=1, num_clients=10, fraction_fit=1)
```
Comment (tanertopal) [inline examples/simulation_pytorch/simulation.py:120]: This is not in sync with the README.md. So I would suggest:

```suggestion
    run_simulation(num_rounds=100, num_clients=10, fraction_fit=0.5)
```

**Review by tanertopal — CHANGES_REQUESTED** [review]
Looks great! Should be ready to merge soon. Left two remarks.

File: examples/simulation_pytorch/README.md (hunk 2 above)
Comment (MakGulati) [reply examples/simulation_pytorch/README.md:36]: @tanertopal Updated them. Thanks and Cheers :)

File: examples/simulation_pytorch/dataset.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    )
+    testloader = torch.utils.data.DataLoader(
+        testset, batch_size=len(testset), shuffle=False, num_workers=2
+    )
+
+    train_X = next(iter(trainloader))[0].numpy()
+    test_X = next(iter(testloader))[0].numpy()
+    train_Y = next(iter(trainloader))[1].numpy()
+    test_Y = next(iter(testloader))[1].numpy()
+    return (train_X, train_Y), (test_X, test_Y)
```
Comment (tanertopal) [inline examples/simulation_pytorch/dataset.py:40]: The model was not converging so I tested a bit extensively. The bug was `next(iter(trainloader))` was called for x and y separately.

This fixes this issue:

```python
    train_xy = next(iter(trainloader))
    test_xy = next(iter(testloader))

    train_x = train_xy[0].numpy()
    train_y = train_xy[1].numpy()

    test_x = test_xy[0].numpy()
    test_y = test_xy[1].numpy()

    return (train_x, train_y), (test_x, test_y)
```

While debugging I also wrote a more PyTorch-only version without NumPy keeping close to your version. I would suggest we take the PyTorch way with `random_splits` as it also loads faster.

This would be the alternative version of `dataset.py`

```python
"""Partitioned version of CIFAR-10 dataset."""

from typing import List, Tuple

import torch
import torchvision
import torchvision.transforms as transforms


def cifar10() -> Tuple[torchvision.datasets.CIFAR10, torchvision.datasets.CIFAR10]:
    """Download dataset from torchvision and convert it to numpy array."""
    transform = transforms.Compose(
        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
    )

    trainset = torchvision.datasets.CIFAR10(
        root="./data", train=True, download=True, transform=transform
    )
    testset = torchvision.datasets.CIFAR10(
        root="./data", train=False, download=True, transform=transform
    )

    return trainset, testset


def create_partitioned_dataloaders(dataset, num_partitions: int) -> List[torch.utils.data.DataLoader]:
    """Create partitioned version of a source dataset."""
    num_per_partition = len(dataset)//num_partitions
    # Seed generator for predictable shuffling of dataset
    generator = torch.Generator().manual_seed(42)
    dataset_partition = torch.utils.data.random_split(dataset, [
        num_per_partition] * num_partitions, generator=generator)
    dataloaders = [torch.utils.data.DataLoader(
        ds_p, batch_size=32, shuffle=True) for ds_p in dataset_partition]
    return dataloaders


def load(num_partitions: int) -> List[Tuple[torch.utils.data.DataLoader, torch.utils.data.DataLoader]]:
    """Create partitioned version of CIFAR-10."""
    trainset, testset = cifar10()
    trainloaders = create_partitioned_dataloaders(trainset, num_partitions)
    testloaders = create_partitioned_dataloaders(testset, num_partitions)
    list_of_dataloaders = [(train_dl, test_dl) for train_dl,
                           test_dl in zip(trainloaders, testloaders)]
    return list_of_dataloaders
```

File: examples/simulation_pytorch/simulation.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            net.load_state_dict(state_dict, strict=True)
+
+        def fit(self, parameters, config):
+            self.set_parameters(parameters)
+            train(net, trainloader, epochs=1)
+            return self.get_parameters(), len(trainloader), {}
+
+        def evaluate(self, parameters, config):
+            self.set_parameters(parameters)
+            loss, accuracy = test(net, testloader)
```
Comment (tanertopal) [inline examples/simulation_pytorch/simulation.py:51]: FYI
```
            loss, accuracy = test(net, testloader)
            print(loss, accuracy)
```
used this for debugging

**Review by tanertopal — CHANGES_REQUESTED** [review]
@MakGulati I tried running the example a little longer to check if it converges. Unfortunately, it did not yet. Left some comments.

I think we are close to the finish line!

File: examples/simulation_pytorch/dataset.py (hunk 4 above)
Comment (MakGulati) [reply examples/simulation_pytorch/dataset.py:40]: @tanertopal Thanks a lot for finding the bug. I am sorry I might have overlooked this part.

Regarding the pytorch only solution, it is a creative idea and  I really liked it. One more thing, I found this issue with random split [open issue](https://github.com/PyTorchLightning/lightning-bolts/issues/312) which says the split doesn't assure balanced class labels. Having said that it is not a significant problem for us as we expect Non-IID distribution anyways.

Comment (tanertopal) [reply examples/simulation_pytorch/dataset.py:40]: @MakGulati Cool that you found the issue.  It should be fine as in the examples, it's not the goal to provide ideal solutions for targeted data distributions but rather give a starting point for the user. I realized you can further simplify the dataset loading in your version with:

```python
def cifar10():
    """Download dataset from torchvision and convert it to numpy array."""
    transform = transforms.Compose(
        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
    )

    trainset = torchvision.datasets.CIFAR10(
        root="./data", train=True, download=True, transform=transform
    )
    testset = torchvision.datasets.CIFAR10(
        root="./data", train=False, download=True, transform=transform
    )
 
    train_xy = trainset.data, np.array(trainset.targets)
    test_xy = testset.data, np.array(testset.targets)

    return train_xy, test_xy
```

I am by no mean a PyTorch expert so not sure if one would use it that way.

Feel free to commit your preferred solution. If you go with the NumPy solution you will have to shuffle the dataset in the create_partitions function as in: https://github.com/adap/flower/blob/main/examples/simulation/dataset.py#L44

Comment (MakGulati) [reply examples/simulation_pytorch/dataset.py]: @tanertopal I think it should be done this time. Please let me know otherwise. I learnt a lot of things from this PR. :)

File: examples/simulation_pytorch/dataset.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import torchvision.transforms as transforms
+from torch.utils.data import DataLoader, TensorDataset
+from numpy import ndarray
+
+XY = Tuple[np.ndarray, np.ndarray]
+XYList = List[XY]
+PartitionedDataset = List[Tuple[XY, XY]]
+
+
+def cifar_to_numpy() -> Tuple[Tuple[ndarray, ndarray], Tuple[ndarray, ndarray]]:
```
Comment (tanertopal) [inline examples/simulation_pytorch/dataset.py:17]:
```suggestion
def cifar_to_numpy() -> Tuple[XY, XY]:
```

File: examples/simulation_pytorch/dataset.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+"""Partitioned version of CIFAR-10 dataset."""
+
+from typing import List, Tuple, cast
+
+import numpy as np
+import torch
+import torchvision
+import torchvision.transforms as transforms
+from torch.utils.data import DataLoader, TensorDataset
+from numpy import ndarray
```
Comment (tanertopal) [inline examples/simulation_pytorch/dataset.py:10]:
```suggestion
```

**Review by tanertopal — COMMENTED** [review]
Left a small suggestion. When applied we can merge it! Thanks for the great work!

File: examples/simulation_pytorch/dataset.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def partition(x: np.ndarray, y: np.ndarray, num_partitions: int) -> XYList:
+    """Split x and y into a number of partitions."""
+    return list(
+        zip(np.array_split(x, num_partitions), np.array_split(y, num_partitions))
+    )
+
+
+def create_partitions(source_dataset: XY, num_partitions: int,) -> XYList:
```
Comment (tanertopal) [inline examples/simulation_pytorch/dataset.py:50]:
```suggestion
def create_partitions(source_dataset: XY, num_partitions: int) -> XYList:
```

File: examples/simulation_pytorch/dataset.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    """Create partitioned version of a source dataset."""
+    x, y = source_dataset
+    x, y = shuffle(x, y)
+    xy_partitions = partition(x, y, num_partitions)
+
+    return xy_partitions
+
+
+def load(num_partitions: int,) -> PartitionedDataset:
```
Comment (tanertopal) [inline examples/simulation_pytorch/dataset.py:59]:
```suggestion
def load(num_partitions: int) -> PartitionedDataset:
```

File: examples/simulation_pytorch/dataset.py (hunk 6 above)
Comment (MakGulati) [reply examples/simulation_pytorch/dataset.py:17]: A side note do you generate type hints manually or use some automated tool. Thanks. @tanertopal

File: examples/simulation_pytorch/dataset.py (hunk 4 above)
Comment (tanertopal) [reply examples/simulation_pytorch/dataset.py]: Happy to hear!

File: examples/simulation_pytorch/dataset.py (hunk 7 above)
Comment (tanertopal) [reply examples/simulation_pytorch/dataset.py:10]: @MakGulati with the type fixes I think this import is not needed anymore

**Review by tanertopal — APPROVED** [review]
Approved! Thank you for the great contribution!
